stablehlo/dialect/Base.cpp

```diff
@@ -176,5 +175,126 @@ Attribute boundsToEncoding(Attribute prototype, ArrayRef<int64_t> bounds) {
+//  c5:  ?, B           ?, C            ?, B+C
+std::pair<int64_t, int64_t> inferConcatenatedDimAndBound(int64_t leftSize,
+                                                         int64_t rightSize,
+                                                         int64_t leftBound,
+                                                         int64_t rightBound) {
+  bool isLeftStaticDim = !isDynamicDimSize(leftSize);
+  bool isRightStaticDim = !isDynamicDimSize(rightSize);
+  int64_t size = ShapedType::kDynamicSize;
+  int64_t bound = ShapedType::kDynamicSize;
+
+  if (isLeftStaticDim && isRightStaticDim) {
+    size = leftSize + rightSize;
+  } else {
+    int64_t leftSizeOrBound = isLeftStaticDim ? leftSize : leftBound;
+    int64_t rightSizeOrBound = isRightStaticDim ? rightSize : rightBound;
+    if (!isDynamicDimSize(leftSizeOrBound) &&
+        !isDynamicDimSize(rightSizeOrBound))
+      bound = leftSizeOrBound + rightSizeOrBound;
+  }
+  return {size, bound};
+}
+
+// Inference rules to merge dimensions with bounds (lhs/rhs are commutative):
+//       Dim of lhs     Dim of rhs      Infer
+//  c0:  X              X               X
+//  c1:  X              ?               X
+//  c2:  X              ?, B(>=X)       X
+//  c3:  X              ?, B(<X)        Will error out by compatible checks
+//  c4:  ?              ?               ?
+//  c5:  ?              ?, B            ?, B
+//  c6:  ?, B           ?, C            ?, min(B, C)
+FailureOr<std::pair<int64_t, int64_t>> inferMergedDimAndBound(
+    Optional<Location> location, int64_t dim, int64_t leftSize,
+    int64_t rightSize, int64_t leftBound, int64_t rightBound) {
+  bool isLeftStaticDim = !isDynamicDimSize(leftSize);
+  bool isRightStaticDim = !isDynamicDimSize(rightSize);
+  bool isLeftStaticBound = !isDynamicDimSize(leftBound);
+  bool isRightStaticBound = !isDynamicDimSize(rightBound);
+  int64_t size = ShapedType::kDynamicSize;
+  int64_t bound = ShapedType::kDynamicSize;
+
+  if (isLeftStaticDim || isRightStaticDim) {
+    if (isLeftStaticDim && isRightStaticDim && leftSize != rightSize)
+      return emitOptionalError(location, "Mismatched dimension sizes ",
+                               leftSize, " and ", rightSize, " in dimension ",
+                               dim);
+    size = isLeftStaticDim ? leftSize : rightSize;
+    if (isLeftStaticBound || isRightStaticBound) {
+      int64_t check_bound = isLeftStaticBound ? leftBound : rightBound;
+      if (size > check_bound)
+        return emitOptionalError(location, "Mismatched dimension size ", size,
+                                 " and bound ", check_bound, " in dimension ",
+                                 dim);
+    }
+  } else {
+    if (isLeftStaticBound && isRightStaticBound)
+      bound = std::min(leftBound, rightBound);
+    else
+      bound = isLeftStaticBound ? leftBound : rightBound;
+  }
+  return std::make_pair(size, bound);
+}
+
+// TODO(zhouxin) Refactor to better handle errors and return single type
```

Comment:
Let's open a ticket about this TODO, so that we don't forget what it means, perhaps with a reference to this PR.

Reply:
Sure opened https://github.com/openxla/stablehlo/issues/408